### PR TITLE
add script for build release

### PR DESCRIPTION
**Pull request by gy95**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add script for build release
And for simplicity, not taking cross platform in consideration, make release only support compile the release of the local operating system.
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -355,3 +355,44 @@ install: _output/local/bin
...
+else ifeq ($(WHAT),edgesite)
+release:
+	KUBEEDGE_OUTPUT_SUBPATH=$(OUT_DIR) hack/make-rules/build.sh edgesite-server edgesite-agent
+	hack/make-rules/release.sh "edgesite" $(VERSION) $(GOARM)
+else ifeq ($(WHAT),keadm)
+release:
+	KUBEEDGE_OUTPUT_SUBPATH=$(OUT_DIR) hack/make-rules/build.sh keadm
+	hack/make-rules/release.sh "keadm" $(VERSION) $(GOARM)
+else ifeq ($(WHAT),kubeedge)
+release:
```
[the remark this answers is not in the document]
Comment (gy95) [reply Makefile:392]: ok, added.

[the remark this answers is not in the document]
Comment (fisherxu) [reply Makefile:392]: Should the version got from the branch?

[the remark this answers is not in the document]
Comment (gy95) [reply Makefile:392]: by default the version comes from git commit info by running `cloudcore --version`, and got like this `1.9.0-54+14e4f891a66a05`, coming from this is due to that want to compile different releases based on the latest codes but not just the same branch.
and we can also specify the version by `make release WHAT=kubeedge VERSION=v1.9.0` for really release.

[the remark this answers is not in the document]
Comment (gy95) [reply Makefile:392]: both the two are almost the same: running `cloudcore --version` also got the branch info, https://github.com/kubeedge/kubeedge/blob/6f5dc3471c12a1cc1973ffeaaa0fe0aceab26d19/pkg/version/version.go#L29-L30

[the remark this answers is not in the document]
Comment (fisherxu) [reply Makefile:392]: It should not specify the version when running make release, from cloudcore or git branch all better, I'd suggest integrating the make in make release, how do you think? @gy95 @zhu733756

File: Makefile [hunk 2]
```diff
@@ -355,3 +356,44 @@ install: _output/local/bin
...
+ifeq ($(HELP),y)
+release:
+	@echo "$$RELEASE_HELP_INFO"
+else ifeq ($(WHAT),edgesite)
+release:
+	KUBEEDGE_OUTPUT_SUBPATH=$(OUT_DIR) hack/make-rules/build.sh edgesite-server edgesite-agent
+	hack/make-rules/release.sh "edgesite" $(VERSION) $(GOARM)
+else ifeq ($(WHAT),keadm)
+release:
+	KUBEEDGE_OUTPUT_SUBPATH=$(OUT_DIR) hack/make-rules/build.sh keadm
```
Comment (gy95) [inline Makefile:389]: I integrate with `make all` cmd before make release, but I'm confused about how the binary version get v1.9.0 but not 1.9.0-54+14e4f891a66a05, So if `cloudcore --version` can get the correct version `v1.9.0`, then the release version will also be the same with binary version `v1.9.0`

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhu733756) [reply Makefile:392]: Every time released, the makefile will patch the default version info like VERSION=v1.9.1?

[the remark this answers is not in the document]
Comment (gy95) [reply Makefile:392]: I haven't understand how version is generated, But I guess that if there's a tag, release use it as the version. And by test, If I checkout a branch and create a tag on it, and then the output is as follows, and release version is also using the tag.
```
# _output/local/bin/cloudcore --version
KubeEdge v1.9.2
```
```
# make release 
KUBEEDGE_OUTPUT_SUBPATH=_output/local hack/make-rules/build.sh
building github.com/kubeedge/kubeedge/cloud/cmd/cloudcore
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/cloudcore -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/cloud/cmd/cloudcore
+ set +x
building github.com/kubeedge/kubeedge/cloud/cmd/admission
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/admission -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/cloud/cmd/admission
+ set +x
building github.com/kubeedge/kubeedge/keadm/cmd/keadm
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/keadm -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/keadm/cmd/keadm
+ set +x
building github.com/kubeedge/kubeedge/edge/cmd/edgecore
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/edgecore -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/edge/cmd/edgecore
+ set +x
building github.com/kubeedge/kubeedge/edgesite/cmd/edgesite-agent
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/edgesite-agent -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/edgesite/cmd/edgesite-agent
+ set +x
building github.com/kubeedge/kubeedge/edgesite/cmd/edgesite-server
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/edgesite-server -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/edgesite/cmd/edgesite-server
+ set +x
building github.com/kubeedge/kubeedge/cloud/cmd/csidriver
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/csidriver -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/cloud/cmd/csidriver
+ set +x
building github.com/kubeedge/kubeedge/cloud/cmd/iptablesmanager
+ go build -o /root/go/src/github.com/kubeedge/kubeedge/_output/local/bin/iptablesmanager -gcflags= -ldflags '-s -w -buildid= -X github.com/kubeedge/kubeedge/pkg/version.buildDate=2021-12-23T07:24:51Z -X github.com/kubeedge/kubeedge/pkg/version.gitCommit=3c28f55841bb402364f69e1629d671f8398bd98f -X github.com/kubeedge/kubeedge/pkg/version.gitTreeState=clean -X github.com/kubeedge/kubeedge/pkg/version.gitVersion=v1.9.2 -X github.com/kubeedge/kubeedge/pkg/version.gitMajor=1 -X github.com/kubeedge/kubeedge/pkg/version.gitMinor=9' github.com/kubeedge/kubeedge/cloud/cmd/iptablesmanager
+ set +x
hack/make-rules/release.sh "kubeedge" "edgesite" "keadm"  
build kubeedge release: v1.9.2 ARCH:amd64
kubeedge-v1.9.2-linux-amd64/
kubeedge-v1.9.2-linux-amd64/cloud/
kubeedge-v1.9.2-linux-amd64/cloud/csidriver/
kubeedge-v1.9.2-linux-amd64/cloud/csidriver/csidriver
kubeedge-v1.9.2-linux-amd64/cloud/iptablesmanager/
kubeedge-v1.9.2-linux-amd64/cloud/iptablesmanager/iptablesmanager
kubeedge-v1.9.2-linux-amd64/cloud/admission/
kubeedge-v1.9.2-linux-amd64/cloud/admission/admission
kubeedge-v1.9.2-linux-amd64/cloud/cloudcore/
kubeedge-v1.9.2-linux-amd64/cloud/cloudcore/cloudcore
kubeedge-v1.9.2-linux-amd64/version
kubeedge-v1.9.2-linux-amd64/edge/
kubeedge-v1.9.2-linux-amd64/edge/edgecore
build edgesite release:v1.9.2 ARCH:amd64
edgesite-v1.9.2-linux-amd64/
edgesite-v1.9.2-linux-amd64/edgesite/
edgesite-v1.9.2-linux-amd64/edgesite/edgesite-server
edgesite-v1.9.2-linux-amd64/edgesite/edgesite-agent
edgesite-v1.9.2-linux-amd64/version
build keadm release: v1.9.2 ARCH:amd64
keadm-v1.9.2-linux-amd64/
keadm-v1.9.2-linux-amd64/keadm/
keadm-v1.9.2-linux-amd64/keadm/keadm
keadm-v1.9.2-linux-amd64/version
```

[the remark this answers is not in the document]
Comment (zhu733756) [reply Makefile:392]: Seems that the GO_LDFLAGS changes as the git tag:
image

See below in golang.sh :
```
 GIT_COMMIT=$(git rev-parse "HEAD^{commit}" 2>/dev/null)

  if git_status=$(git status --porcelain 2>/dev/null) && [[ -z ${git_status} ]]; then
    GIT_TREE_STATE="clean"
  else
    GIT_TREE_STATE="dirty"
  fi

  GIT_VERSION=$(git describe --tags --abbrev=14 "${GIT_COMMIT}^{commit}" 2>/dev/null)
```

[the remark this answers is not in the document]
Comment (zhu733756) [reply Makefile:392]: I confirm it comes from the git tag.

[the remark this answers is not in the document]
Comment (gy95) [reply Makefile:392]: So it seems that we don't need to specify the VERSION like `make release VERSION=v1.9.0` anymore, need I delete this parameter or just keep it left. @zhu733756 @fisherxu

[the remark this answers is not in the document]
Comment (fisherxu) [reply Makefile:392]: > I confirm it comes from the git tag.

Yes, the version is from tag.

[the remark this answers is not in the document]
Comment (fisherxu) [reply Makefile:392]: The version parameter should be deleted.

File: Makefile [hunk 3]
```diff
@@ -351,3 +351,39 @@ install: _output/local/bin
...
+	hack/make-rules/release.sh "keadm" $(GOARM)
+else ifeq ($(WHAT),kubeedge)
+release:
+	KUBEEDGE_OUTPUT_SUBPATH=$(OUT_DIR) hack/make-rules/build.sh cloudcore admission edgecore csidriver iptablesmanager
+	hack/make-rules/release.sh "kubeedge" $(GOARM)
+else
+release:
+	KUBEEDGE_OUTPUT_SUBPATH=$(OUT_DIR) hack/make-rules/build.sh
+	hack/make-rules/release.sh "kubeedge" "edgesite" "keadm" $(GOARM)
+endif
```
Comment (fisherxu) [inline Makefile:389]: It's better to move the if else logic to the release script :)

Comment (gy95) [reply Makefile]: fixed.

File: Makefile [hunk 4]
```diff
@@ -351,3 +351,26 @@ install: _output/local/bin
           done ; \
         fi
 endif
+
+define RELEASE_HELP_INFO
+# release components.
+#
+# Args:
+#   WHAT: Component names to be release. support: kubeedge/edgesite/keadm
+#         If not specified, "everything" will be build and release.
```
Comment (fisherxu) [inline Makefile:360]:
```suggestion
#   WHAT: Component names to be released. Support: kubeedge/edgesite/keadm
#         If not specified, "everything" will be built and released.
```

File: Makefile [hunk 5]
```diff
@@ -351,3 +351,26 @@ install: _output/local/bin
...
+#
+# Example:
+#   make release
+#   make release HELP=y
+#   make release WHAT=kubeedge/edgesite/keadm
+#   make release WHAT=kubeedge GOARM=GOARM7/GOARM8
```
Comment (fisherxu) [inline Makefile:366]: I think `\` will not be supported? If not supported, it's better not write in the help info :)
```
make release WHAT=kubeedge GOARM=GOARM7
make release WHAT=kubeedge GOARM=GOARM8
```

File: hack/make-rules/release.sh [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+      ARCH="arm64"
+    else
+      targets+=("${arg}")
+    fi
+  done
+
+  if [[ ${#targets[@]} -eq 0 ]]; then
+    targets+=("kubeedge")
+    targets+=("edgesite")
+    targets+=("keadm")
```
Comment (fisherxu) [inline hack/make-rules/release.sh:45]: Define a `ALL_RELEASE_TARGETS` value,  then `targets=("${ALL_RELEASE_TARGETS[@]}")` like in make build.

File: hack/make-rules/release.sh [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  fi
+
+  kubeedge::version::get_version_info
+  VERSION=${GIT_VERSION} #got v1.9.0-54+14e4f891a66a05
```
Comment (fisherxu) [inline hack/make-rules/release.sh:49]: Remove  the comment.

File: hack/make-rules/release.sh [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+  for bin in ${targets[@]}; do
+    if [ "${bin}" == "edgesite" ]; then
+      hack/make-rules/build.sh edgesite-server edgesite-agent
+      echo "build edgesite release:"$VERSION "ARCH:"$ARCH
+      build_edgesite_release $VERSION $ARCH
```
Comment (fisherxu) [inline hack/make-rules/release.sh:55]: Can init the VERSION and print the log in `build_edgesite_release`, same in below

File: hack/make-rules/release.sh [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+      echo "build kubeedge release:" $VERSION "ARCH:"$ARCH
+      build_kubeedge_release $VERSION $ARCH
+    else
+      echo "not supported release: " $bin
+      exit 1
+    fi
+  done
+}
+
+function build_kubeedge_release() {
```
Comment (fisherxu) [inline hack/make-rules/release.sh:71]: It's better to compile the binary before make the release.

File: hack/make-rules/release.sh (hunk 8 above)
Comment (gy95) [reply hack/make-rules/release.sh]: done

File: hack/make-rules/release.sh (hunk 9 above)
Comment (gy95) [reply hack/make-rules/release.sh:86]: need to build different binaries according to the different release names.

File: hack/make-rules/release.sh [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+  kubeedge::version::get_version_info
+  VERSION=${GIT_VERSION}
+
+  for bin in ${targets[@]}; do
+    if [ "${bin}" == "edgesite" ]; then
+      if [ "${ARCH}" == "arm" ] || [ "${ARCH}" == "arm" ]; then
+        hack/make-rules/crossbuild.sh edgesite-server edgesite-agent $(GOARM)
+      else
+        hack/make-rules/build.sh edgesite-server edgesite-agent
+      fi
```
Comment (fisherxu) [inline hack/make-rules/release.sh:60]:
```suggestion
      if [ "${ARCH}" == "amd64" ]; then
        hack/make-rules/build.sh edgesite-server edgesite-agent
      else
        hack/make-rules/crossbuild.sh edgesite-server edgesite-agent $(GOARM)
      fi
```

Comment (gy95) [reply hack/make-rules/release.sh]: done

File: hack/make-rules/release.sh [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+  echo ${VERSION} > _output/local/kubeedge-${VERSION}-linux-${ARCH}/version
+  cp _output/local/bin/admission _output/local/kubeedge-${VERSION}-linux-${ARCH}/cloud/admission
+  cp _output/local/bin/cloudcore _output/local/kubeedge-${VERSION}-linux-${ARCH}/cloud/cloudcore
+  cp _output/local/bin/csidriver _output/local/kubeedge-${VERSION}-linux-${ARCH}/cloud/csidriver
+  cp _output/local/bin/iptablesmanager _output/local/kubeedge-${VERSION}-linux-${ARCH}/cloud/iptablesmanager
+
+  cp _output/local/bin/edgecore _output/local/kubeedge-${VERSION}-linux-${ARCH}/edge
+
+  cd _output/local
+  tar -czvf ${KUBEEDGE_ROOT}/_output/local/kubeedge-${VERSION}-linux-${ARCH}.tar.gz kubeedge-${VERSION}-linux-${ARCH}/
```
Comment (fisherxu) [inline hack/make-rules/release.sh:120]: Just have a try, run as expected! How about create a `release` dir under `local`, and also create a subdir named as version like `v1.9.0`, then move all the release package under it?

**Review by fisherxu — COMMENTED** [review]
/approve
Squash the commits.

**Comment by fisherxu** [comment]
cc @zhu733756 for final review.

**Comment by zhu733756** [comment]
> cc @zhu733756 for final review.

LGTM

**Comment by gy95** [comment]
ping @fisherxu need LGTM

**Pull request merged**